### ISSUE TITLE
Request Timeouts and Protocol Errors occurring with HAProxy 2.9

### ISSUE THREAD

**Issue by gygitlab** [haproxy/haproxy#2387]
### Detailed Description of the Problem

This is a bit of a weird one frankly but I've confirmed this to be an issue that looks to appear in HAProxy 2.9 only.

We run performance tests against our application with a HAProxy load balancer in front, deployed via the docker image. With 2.9 we started noticing some tests having a small number of failed requests that weren't occurring before.

Digging deeper the failed requests were actually request timeouts but with no meaningful error logs to suggest why, the logs actually just contained the full page code for some reason.

Digging further we noticed this only seemed to happen on requests to endpoints that return JSON files and we also noticed a small number of vague `connection error: PROTOCOL_ERROR` logs in our client (although this error specifically only seems to happen with the alpine image, unsure if that's a red herring).

Unfortunately I've not got much else to go on, it's just super weird, but I can say through process of elimination that this only occurs with HAProxy 2.9 with 2.8.5 working completely as normal.

### Expected Behavior

That requests succeed with no timeouts or protocol errors

### Steps to Reproduce the Behavior

Not sure if I can provide a straightforward recreation except perhaps to have a setup where HAProxy is serving JSON and to try calling that en masse?

### Do you have any idea what may have caused this?

I saw there was a reference to JSON in the release notes but this seems like an optional feature. I fear this may be something more guttural.

### What is your configuration?

```haproxy
global
    log /dev/log local0
    log localhost local1 notice
    log stdout format raw local0
    maxconn 10000
    daemon

defaults
    log global
    mode http
    default-server inter 10s fall 3 rise 2
    balance leastconn
    compression algo gzip
    compression type text/css text/html text/plain text/xml application/json
    option httplog
    option dontlognull
    retries 3
    timeout connect 30s
    timeout client 305s
    timeout server 305s

frontend http-in
    bind *:80
    bind *:443 ssl crt /usr/local/etc/haproxy/ssl/ strict-sni

    http-request set-header X-Forwarded-Proto https if { ssl_fc }
    http-request set-header X-Forwarded-Ssl on if { ssl_fc }

    acl letsencrypt path_beg /.well-known/acme-challenge/
    use_backend letsencrypt if letsencrypt

    default_backend rails

backend rails
    option httpchk GET /-/readiness
    option forwardfor
    redirect scheme https if !{ ssl_fc }

    server rails1 <ip_redacted> check inter 3s fall 1
    server rails2 <ip_redacted> check inter 3s fall 1
    server rails3 <ip_redacted> check inter 3s fall 1

backend letsencrypt
    server acme-sh host.docker.internal:88

listen stats
    bind *:1936
    http-request use-service prometheus-exporter if { path /metrics }
    mode http
    stats enable
    stats uri /
    stats refresh 10s
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.9.0-fddb8c1 2023/12/05 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: http://www.haproxy.org/bugs/bugs-2.9.0.html
Running on: Linux 5.15.0-1053-azure #61~20.04.1-Ubuntu SMP Tue Nov 21 17:50:57 UTC 2023 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 1.1.1w  11 Sep 2023
Running on OpenSSL version : OpenSSL 1.1.1w  11 Sep 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

### Last Outputs and Backtraces

```plain
I couldn't find any consistent error logs from HAProxy, but I did sometimes see the following but unsure if this was due to the client closing connection due to timeout:


<ip_redacted> [12/Dec/2023:09:54:08.774] http-in rails/<NOSRV> 0/-1/-1/-1/0 302 102 - - LR-- 24/23/0/0/0 0/0 "GET / HTTP/1.1"
<ip_redacted> [12/Dec/2023:09:54:09.627] http-in/2: SSL handshake failure (error:0A0000EA:SSL routines::callback failed)
```
```


### Additional Information

_No response_

**Comment by rtammekivi** [comment]
Seeing the same issues. ERR_HTTP2_PROTOCOL_ERROR reported on chrome side, with SD on the server side.

**Comment by capflam** [comment]
@gygitlab, I'm wondering if your issue could be related to #2382. @Blefish, in your case, it could also be related to #2388.

**Comment by capflam** [comment]
@gygitlab, @Blefish, we want to emit a 2.9.1 tomorrow. Have you any way to confirm your issue is fixed by testing the 3.0-DEV ?

**Comment by gygitlab** [comment]
@capflam I'll see if I can test this today thanks.

**Comment by capflam** [comment]
Cool, many thanks !

**Comment by rtammekivi** [comment]
I built a new docker image from https://github.com/docker-library/haproxy/blob/master/3.0/Dockerfile against
`ENV HAPROXY_URL https://www.haproxy.org/download/3.0/src/snapshot/haproxy-ss-LATEST.tar.gz`

```
haproxy -v
HAProxy version 3.0-dev0-682f73b 2023/12/13 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 6.1.52-71.125.amzn2023.aarch64 #1 SMP Tue Sep 12 21:41:10 UTC 2023 aarch64
```

Seems to be working fine :+1:

**Comment by capflam** [comment]
Good news, thanks !

**Comment by mitko9922** [comment]
And I want to tell about a problem with wordpress with version 2.9 docker container, can't get into the backend(wp-admin) of wordpress the error is as follows:
Failed to load resource: net::ERR_QUIC_PROTOCOL_ERROR
SO resource  zxcvbn.min.js not loaded.
If return the container with the previous haproxy version everything works as it should!

**Comment by capflam** [comment]
It don't know if it is the same issue because yours is a QUIC error. However several bugs were recently fixed. Some about QUIC. A 2.9.1 will be emitted soon. In the mean time, if possible, you can test HEAD of 3.0-dev to check if it fixes your issue.

**Comment by gygitlab** [comment]
Hey, apologies for the delay but just to confirm the issue has indeed be resolved with 2.9.1. Thanks!

**Comment by capflam** [comment]
Thanks for the feedback. I'm closing the issue.
